tidalapi/page.py

```diff
@@ -19,54 +19,100 @@
 """
 
 import copy
-from typing import TYPE_CHECKING, Any, Dict, Iterator, List, Optional, Union, cast
+from dataclasses import dataclass
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Union,
+    cast,
+)
 
 from tidalapi.types import JsonObj
 
 if TYPE_CHECKING:
-    import tidalapi
+    from tidalapi.album import Album
+    from tidalapi.artist import Artist
+    from tidalapi.media import Track, Video
+    from tidalapi.mix import Mix
+    from tidalapi.playlist import Playlist, UserPlaylist
+    from tidalapi.request import Requests
+    from tidalapi.session import Session
 
+PageCategories = Union[
+    "Album",
+    "PageLinks",
+    "FeaturedItems",
+    "ItemList",
+    "TextBlock",
+    "LinkList",
+    "Mix",
+]
 
-class Page(object):
+AllCategories = Union["Artist", PageCategories]
+
+
+def _no_op_items(*args) -> List[Any]:
+    """A simple passthrough stub for category types that are not iterable."""
+    return []
+
+
+class Page:
     """
     A page from the https://listen.tidal.com/view/pages/ endpoint
 
     The :class:`categories` field will the most complete information
     However it is an iterable that goes through all the visible items on the page as well, in the natural reading order
     """
 
-    title = ""
-    categories: Optional[List[Any]] = None
-    _categories_iter: Optional[Iterator[Any]] = None
+    title: Optional[str] = ""
+    categories: Optional[List["AllCategories"]] = None
+    _categories_iter: Optional[Iterator["AllCategories"]] = None
+    _items_iter: Optional[Iterator[Callable[..., Any]]] = None
+    page_category: "PageCategory"
+    request: "Requests"
 
-    def __init__(self, session, title):
+    def __init__(self, session: "Session", title: Optional[str]):
         self.request = session.request
         self.categories = None
         self.title = title
         self.page_category = PageCategory(session)
 
-    def __iter__(self):
+    def __iter__(self) -> "Page":
```

Comment:
(aside: when we move to 3.11 these should become `typing.Self` to handle children.)